File: src/input.cpp
#include <raylib.h>
#include <string>
#include <vector>
#include <iostream>
// #include <winsock2.h>
#define FONT_SIZE 17
#define PADDINGX 5
#define PADDINGY 8
#define MAX_LINEAS 3


/*
MEJORAR ESTA CLASE
LE FALTA MEJORAR LA FUENTE
MEJORAR TEMA DE LEER TECLAS ESPECIALES COMO ? ! #, ETC
PERMITIR USO DE FLECHAS PARA MOVERSE POR EL TEXTO
MANTENER APRETADA UNA TECLA ES COMO TOCARLA VARIAS VECES
SI MANTENGO DELETE DEBERIA BORRAR HASTA QUE DEJE DE APRETARLO
*/ 
class InputField 
{
    public:
        InputField(int x, int y, int ancho, int alto) : text("") 
        {
            campo.x=x;
            campo.y=y;
            campo.height=alto;
            campo.width=ancho;
            tamLinea=0;
            cantLineas=0;
        }

        void processInput() 
        {   
            int key = GetKeyPressed();
            if(cantLineas>=MAX_LINEAS)
            {
                if (key == KEY_ENTER) 
                {
                    // sendTextOverSocket();
                    borrarCampo();
                }
                return;
            }
            while (key > 0) 
            {
                if (key == KEY_BACKSPACE) 
                {
                    if (!text.empty()) 
                    {
                        text.pop_back(); 
                        tamLinea = tamLinea>0 ? tamLinea--  : 0 ;
                    }
                } else if (key == KEY_ENTER) 
                {
                    // sendTextOverSocket();
                    borrarCampo();
                } else 
                {
                    text += (char)key; 
                    tamLinea++;
                }
                key = GetKeyPressed(); 
            }
        }

        void dibujar() 
        {
            // Dibujar el campo de texto
            Color colorModificado = BLACK; 
            colorModificado.a = (unsigned char)(0.5 * 255); 
            DrawRectangleRec(campo, colorModificado);
            DrawText(text.c_str(), campo.x + PADDINGX, campo.y + PADDINGY , FONT_SIZE, WHITE);
        }
        void dibujar(float crlf) 
        {
            cantLineas++;
            if(cantLineas<=MAX_LINEAS)
            {
                text += '\n'; 
                dibujar();
            }

        }
        bool overflow()
        {   
            if(tamLinea > campo.width/FONT_SIZE + FONT_SIZE - PADDINGX)
            {
                tamLinea=0;
                return true;
            }
            return false;
        }
        void  borrarCampo()
        {
            text.clear(); 
            tamLinea=0;
            cantLineas=0;
        }
        void sendTextOverSocket()
        {
            text.clear(); 
        }
    private:
        Rectangle campo;
        int tamLinea;
        int cantLineas;
        std::string text;
};

File: src/input.py
import pyray as rl
from chatHistory import ChatHistory
FONT_SIZE = 17
PADDING_X = 5
PADDING_Y = 8
MAX_LINEAS = 3


class InputField:

    def __init__(self, x, y, ancho, alto, chat_history , text="" ) -> None:
        self._campo = rl.Rectangle(x, y, ancho, alto)

        self._text = text

        self._tam_linea = 0
        self._cant_lineas = 0
        
        self._chat_history = chat_history

    def process_input(self):
        key = rl.get_key_pressed()
        if self._cant_lineas >= MAX_LINEAS:
            if key == rl.KeyboardKey.KEY_ENTER:
                self.send_text()
            return
        while key > 0:
            if key == rl.KeyboardKey.KEY_BACKSPACE:
                if self._text:
                    self._text = self._text[:-1]
                    self._tam_linea = self._tam_linea - 1 if self._tam_linea > 0 else 0
            elif key == rl.KeyboardKey.KEY_ENTER:
                self.send_text()
            else:
                self._text += chr(key)
                self._tam_linea += 1
            key = rl.get_key_pressed()

    def dibujar(self):
        color_modificado = rl.Color(0, 0, 0, 255)
        color_modificado.a = int(0.5 * 255)
        rl.draw_rectangle_rec(self._campo, color_modificado)
        rl.draw_text(self._text, int(self._campo.x + PADDING_X),
                     int(self._campo.y + PADDING_Y), FONT_SIZE, rl.WHITE)

    def dibujar_con_crlf(self, crlf):
        self._cant_lineas += 1
        if self._cant_lineas <= MAX_LINEAS:
            self._text += '\n'
            self.dibujar()

    def overflow(self) -> bool:
        if self._tam_linea > self._campo.width / FONT_SIZE + FONT_SIZE - PADDING_X:
            self._tam_linea = 0
            return True

        return False

    def borrar_campo(self):
        self._text = ""
        self._tam_linea = 0
        self._cant_lineas = 0

    def send_text(self):
        if (self._chat_history.es_mi_turno()):
            self._chat_history.recive_data_input(self._text)
            self.borrar_campo()
